_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+---
+layout: post
+title: Building a Drawer component in React
+category:
+- tutorials
+tags:
+- React
+- SCSS
+---
+
+In this article, we're going to be building out a React Drawer component. We're going to complete several iterations of a it into something that could be reusable anywhere throughout your application.
+<!--excerpt-->
+
+## What we're going to build
+
+For the purposes of this tutorial, we're going to be constructing a drawer component. The
+end feature will look like this:
+
+![Final Drawer Component](/img/2019/final-drawer.gif)
+
+The first step we'll take is listing out all our high-level feature tasks to complete.
+
+* A clickable title that on click expands an attached content section
+* Content section that animates like a drawer
+* Screen reader compatible tab functionality
+
+With that we can get started on the component.
+
+## Building the basic React Component
+
+The first thing we'll need is something to store the current state of the container which will either be expanded or collapsed. We can use a state variable `active` to denote this in our component's constructor.
+
+```jsx
+import React from 'react';
+
+class Drawer extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      active: false,
+    };
+  }
+}
+
+export default Drawer;
+```
+
+We'll also add in the HTML necessary for the component within its render method. `title` will
+be a prop directly passed into the component like `<Drawer title="Cool title">` while `children` is a special prop that contains all child elements inside of a component. This means using our component like this:
+
+``` jsx
+<Drawer>
+  <p>hi</p>
+</Drawer>
+```
+
+Means that within `Drawer` the `this.props.children` prop returns the inner content of `<p>hi</p>`.
+
+By structuring it in this manner we allow future developers to compose the content they want displayed within the container. It affords a lot of freedom and flexibility in how the component is consumed.
+
+``` jsx
+render() {
+  return (
+    <div className="Drawer">
+      <div className="Drawer-title">
+        {this.props.title}
+      </div>
+
+      <div className="Drawer-content">
+        {this.props.children}
+      </div>
+    </div>
+  );
+}
+```
+
+With a little variable destructing, we can make this even simpler.
+
+``` jsx
+render() {
+  const { title, children } = this.props;
+
+  return (
+    <div className="Drawer">
+      <div className="Drawer-title">
+        {title}
+      </div>
+
+      <div className="Drawer-content">
+        {children}
+      </div>
+    </div>
+  );
+}
+```
+
+Next we need to toggle the above state variable `active` whenever a user clicks the title. We can do this by adding an `onClick` handler to the element.
+
```

Comment:
Give a summary here of what you've done so far. What would the code do if you ran it as is? Then add the next piece